### PR TITLE
Unify single-response-requests with streaming requests on `graphql-transport-ws`

### PR DESCRIPTION
**Pull request by kristjanvalur**
(replaces closed #1739)

Various problems with single-response-requests existed as a result of their having a separate control
path from streaming requests.  Single-response-requests are really just special cases of streaming requests
that deliver __at most__ one `Next` message and need to behave as such.

## Description

single-response-request are now handled as *tasks* and can be overlapped with other requests
on the same connection, can be cancelled, and share the same unique requests ID namespace.

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #1731
* #1733
* #1734
* #1738

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This release fixes a number of problems with single-result-operations over
`graphql-transport-ws` protocol

- operation **IDs** now share the same namespace as streaming operations
  meaning that they cannot be reused while the others are in operation

- single-result-operations now run as *tasks* meaning that messages related
  to them can be overlapped with other messages on the websocket.

- single-result-operations can be cancelled with the `complete` message.

- IDs for single result and streaming result operations are now released
  once the operation is done, allowing them to be re-used later, as well as
  freeing up resources related to previous requests.


---
Here's the preview release card for twitter:




Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @kristjanvalur for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -45,7 +45,7 @@ repos:
       - id: check-toml
 
   - repo: https://github.com/psf/black
-    rev: 22.1.0
+    rev: 22.3.0
```
Comment (estyxx) [inline .pre-commit-config.yaml:48]: Maybe is better to update black in a separate PR and run it on all file as well :)

Comment (kristjanvalur) [reply .pre-commit-config.yaml:48]: To clarify:  Without it, the commit checks would not run.

File: tests/aiohttp/test_graphql_transport_ws.py [hunk 2]
```diff
@@ -285,6 +285,55 @@ async def test_duplicated_operation_ids(aiohttp_client):
         assert data.extra == "Subscriber for sub1 already exists"
 
 
+async def test_reused_operation_ids(aiohttp_client):
```
Comment (BryceBeagle) [inline tests/aiohttp/test_graphql_transport_ws.py:288]: As these are fairly complicated tests, could you add some docstrings describing what functionality is being tested

**Review by BryceBeagle — CHANGES_REQUESTED** [review]

File: tests/aiohttp/test_graphql_transport_ws.py (hunk 2 above)
Comment (kristjanvalur) [reply tests/aiohttp/test_graphql_transport_ws.py:288]: Added docstrings.
It is a bit annoying that we have three identical sets of tests, testing _internal_ functionality, that is in theory quite independent of the integration in use.   Would it make sense to separate "integration" tests from unit tests?

Comment (BryceBeagle) [reply tests/aiohttp/test_graphql_transport_ws.py:288]: cc @patrick91, I think we were discussing ideas for doing this last week

Comment (patrick91) [reply tests/aiohttp/test_graphql_transport_ws.py:288]: @kristjanvalur totally, I've opened up an issue for this: https://github.com/strawberry-graphql/strawberry/issues/1809

File: strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py [hunk 3]
```diff
@@ -204,24 +170,67 @@ async def handle_streaming_operation(self, message: SubscribeMessage) -> None:
...
+            async def get_result_source():
+                yield await self.schema.execute(
+                    query=message.payload.query,
+                    variable_values=message.payload.variables,
+                    context_value=context,
+                    root_value=root_value,
+                    operation_name=message.payload.operationName,
+                )
+
+            result_source = get_result_source()
```
Comment (patrick91) [inline strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py:193]: maybe in a future PR we can move this to a classmethod to clean this method 😊

**Review by patrick91 — APPROVED** [review]
this looks great! thanks so much and sorry for the late review!

**Pull request merged**